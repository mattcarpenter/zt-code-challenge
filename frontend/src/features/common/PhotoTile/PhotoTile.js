import React from 'react';
import PropTypes from 'prop-types';
import { withSize } from 'react-sizeme';

const PhotoTile = ({profileURL, thumbnailURL, originalURL, profileImage, uploaderName, id, originalWidth, originalHeight, canFavorite, size }) => {
  const scaledHeight = calculateScaledHeight(originalWidth, originalHeight, size.width);

  return (
    <div style={{ height: scaledHeight }}>
      <img src={thumbnailURL} width={size.width} height={scaledHeight}/>
    </div>
  );
};

PhotoTile.propTypes = {
  profileURL: PropTypes.string,
  thumbnailURL: PropTypes.string.isRequired,
  originalURL: PropTypes.string,
  id: PropTypes.string.isRequired,
  profileImage: PropTypes.string,
  uploaderName: PropTypes.string,
  originalWidth: PropTypes.number.isRequired,
  originalHeight: PropTypes.number.isRequired,
  canFavorite: PropTypes.bool
};

export default withSize()(PhotoTile);

export const calculateScaledHeight = (originalWidth, originalHeight, newWidth) => {
  return originalHeight * newWidth / originalWidth;
};
