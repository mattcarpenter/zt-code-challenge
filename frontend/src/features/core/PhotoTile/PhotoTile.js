import React from 'react';
import PropTypes from 'prop-types';
import { withSize } from 'react-sizeme';
import { makeStyles } from '@material-ui/core/styles';
import SaveAltIcon from '@material-ui/icons/SaveAlt';
import AddIcon from '@material-ui/icons/Add';
import Button from '@material-ui/core/Button';

const useStyles = makeStyles({
  root: {
    position: 'relative',
    '&:hover > div': {
      display: 'block'
    },
    margin: '12px 6px',
    borderRadius: 3,
    overflow: 'hidden',
    backgroundColor: 'rgb(100,100,100)'
  },
  image: {
    position: 'absolute',
    top: 0,
    left: 0,
    zIndex: 0
  },
  meta: {
    backgroundColor: 'rgba(0,0,0,0.7)',
    position: 'absolute',
    color: '#fff',
    bottom: 0,
    right: 0,
    left: 0,
    zIndex: 1,
    padding: 10
  },
  actions: {
    position: 'absolute',
    top: 15,
    right: 15,
    zIndex: 1
  },
  profileImage: {
    width: 32,
    height: 32,
    borderRadius: 16,
    verticalAlign: 'middle',
    marginRight: 15
  },
  profileLink: {
    textDecoration: 'none',
    color: '#aaa',
    transition: 'color .5s',
    '&:hover': {
      color: '#fff',
      transition: 'color .5s'
    }
  },
  button: {
    marginLeft: 10,
    minWidth: 0,
    padding: '6px 10px',
    '& svg': {
      fill: '#666'
    }
  },
  cover: {
    display: 'none',
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0
  }
});

const PhotoTile = ({ id, thumbnailURL, profileImageURL, profileUserName, originalWidth,
                     originalHeight, canFavorite, size, profileURL, onFavorite, downloadURL }) => {
  const scaledHeight = calculateScaledHeight(originalWidth, originalHeight, size.width);
  const classes = useStyles();

  const handleDownloadClick = () => {
    // tricks the browser into initiating a download for this resource
    var anchor = document.createElement('a');
    anchor.href = downloadURL;
    anchor.target = '_blank';
    anchor.download = id;
    anchor.click();
  };

  return (
    <div style={{ height: scaledHeight }} className={classes.root}>
      <img
        className={classes.image} src={thumbnailURL}
        width={size.width}
        height={scaledHeight} />
      <div className={classes.cover}>
        <div className={classes.meta}>
          <a href={profileURL} target="_BLANK" className={classes.profileLink} rel="noopener noreferrer">
            <img src={profileImageURL} className={classes.profileImage} />
            { profileUserName }
          </a>
        </div>
        <div className={classes.actions}>
          { canFavorite && (
            <Button
              aria-label="Add to Favorites"
              onClick={onFavorite}
              variant="contained"
              className={classes.button}
              color="default">
              <AddIcon />
            </Button>
          )}
          <Button
            aria-label="Download Image"
            onClick={handleDownloadClick}
            variant="contained"
            className={classes.button}
            color="default">
            <SaveAltIcon />
          </Button>
        </div>
      </div>
    </div>
  );
};

PhotoTile.propTypes = {
  id: PropTypes.string.isRequired,
  profileURL: PropTypes.string,
  thumbnailURL: PropTypes.string.isRequired,
  profileImageURL: PropTypes.string,
  downloadURL: PropTypes.string,
  profileUserName: PropTypes.string,
  originalWidth: PropTypes.number.isRequired,
  originalHeight: PropTypes.number.isRequired,
  canFavorite: PropTypes.bool,
  onFavorite: PropTypes.func
};

PhotoTile.defautProps = {
  onFavorite: () => {}
};

export default withSize()(PhotoTile);

export const calculateScaledHeight = (originalWidth, originalHeight, newWidth) => {
  return originalHeight * newWidth / originalWidth;
};
