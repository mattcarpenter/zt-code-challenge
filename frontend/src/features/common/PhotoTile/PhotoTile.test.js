import React from 'react';
import { mount } from 'enzyme';
import PhotoTile, {calculateScaledHeight} from './PhotoTile';

describe('PhotoTile', () => {

  describe('component', () => {

    it('properly renders', () => {
      const wrapper = mount(
        <PhotoTile
          size={{width: 200}}
          originalWidth={2000}
          originalHeight={1500}
          thumbnailURL="abc"
          id="def"
        />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('calls the onFavorite method when the Add to Favorites button is clicked', () => {
      const onFavorite = jest.fn();
      const wrapper = mount(
        <PhotoTile
          size={{width: 200}}
          originalWidth={2000}
          originalHeight={1500}
          onFavorite={onFavorite}
          thumbnailURL="abc"
          id="def"
          canFavorite={true}
        />
      );
      wrapper.find('button').find({'aria-label':'Add to Favorites'}).simulate('click');
      expect(onFavorite.mock.calls.length).toBe(1);
    });

    it('calls the onDownload method when the Add to Favorites button is clicked', () => {
      const onDownload = jest.fn();
      const wrapper = mount(
        <PhotoTile
          size={{width: 200}}
          originalWidth={2000}
          originalHeight={1500}
          onDownload={onDownload}
          thumbnailURL="abc"
          id="def"
          canFavorite={true}
        />
      );
      wrapper.find('button').find({'aria-label':'Download Image'}).simulate('click');
      expect(onDownload.mock.calls.length).toBe(1);
    });

  });

  describe('calculateScaledHeight', () => {

    it('calculates the correct new height given original dimensions and a new width', () => {
      expect(calculateScaledHeight(800, 600, 400)).toBe(300);
      expect(calculateScaledHeight(600, 800, 300)).toBe(400);
      expect(calculateScaledHeight(500,500,250)).toBe(250);
    });

  });

});
