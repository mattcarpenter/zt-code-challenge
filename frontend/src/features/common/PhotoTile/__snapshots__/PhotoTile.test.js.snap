// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoTile component properly renders 1`] = `
<SizeMe(PhotoTile)
  id="def"
  originalHeight={1500}
  originalWidth={2000}
  size={
    Object {
      "width": 200,
    }
  }
  thumbnailURL="abc"
>
  <SizeMeRenderer(PhotoTile)
    disablePlaceholder={false}
    explicitRef={[Function]}
    id="def"
    originalHeight={1500}
    originalWidth={2000}
    size={
      Object {
        "width": 200,
      }
    }
    thumbnailURL="abc"
  >
    <SizeMeReferenceWrapper>
      <PhotoTile
        id="def"
        originalHeight={1500}
        originalWidth={2000}
        size={
          Object {
            "width": 200,
          }
        }
        thumbnailURL="abc"
      >
        <div
          className="makeStyles-root-1"
          style={
            Object {
              "height": 150,
            }
          }
        >
          <img
            className="makeStyles-image-2"
            height={150}
            src="abc"
            width={200}
          />
          <div
            className="makeStyles-cover-8"
          >
            <div
              className="makeStyles-meta-3"
            >
              <a
                className="makeStyles-profileLink-6"
                rel="noopener noreferrer"
                target="_BLANK"
              >
                <img
                  className="makeStyles-profileImage-5"
                />
              </a>
            </div>
            <div
              className="makeStyles-actions-4"
            >
              <WithStyles(ForwardRef(Button))
                aria-label="Download Image"
                className="makeStyles-button-7"
                color="default"
                variant="contained"
              >
                <ForwardRef(Button)
                  aria-label="Download Image"
                  className="makeStyles-button-7"
                  classes={
                    Object {
                      "colorInherit": "MuiButton-colorInherit",
                      "contained": "MuiButton-contained",
                      "containedPrimary": "MuiButton-containedPrimary",
                      "containedSecondary": "MuiButton-containedSecondary",
                      "containedSizeLarge": "MuiButton-containedSizeLarge",
                      "containedSizeSmall": "MuiButton-containedSizeSmall",
                      "disableElevation": "MuiButton-disableElevation",
                      "disabled": "Mui-disabled",
                      "endIcon": "MuiButton-endIcon",
                      "focusVisible": "Mui-focusVisible",
                      "fullWidth": "MuiButton-fullWidth",
                      "iconSizeLarge": "MuiButton-iconSizeLarge",
                      "iconSizeMedium": "MuiButton-iconSizeMedium",
                      "iconSizeSmall": "MuiButton-iconSizeSmall",
                      "label": "MuiButton-label",
                      "outlined": "MuiButton-outlined",
                      "outlinedPrimary": "MuiButton-outlinedPrimary",
                      "outlinedSecondary": "MuiButton-outlinedSecondary",
                      "outlinedSizeLarge": "MuiButton-outlinedSizeLarge",
                      "outlinedSizeSmall": "MuiButton-outlinedSizeSmall",
                      "root": "MuiButton-root",
                      "sizeLarge": "MuiButton-sizeLarge",
                      "sizeSmall": "MuiButton-sizeSmall",
                      "startIcon": "MuiButton-startIcon",
                      "text": "MuiButton-text",
                      "textPrimary": "MuiButton-textPrimary",
                      "textSecondary": "MuiButton-textSecondary",
                      "textSizeLarge": "MuiButton-textSizeLarge",
                      "textSizeSmall": "MuiButton-textSizeSmall",
                    }
                  }
                  color="default"
                  variant="contained"
                >
                  <WithStyles(ForwardRef(ButtonBase))
                    aria-label="Download Image"
                    className="MuiButton-root MuiButton-contained makeStyles-button-7"
                    component="button"
                    disabled={false}
                    focusRipple={true}
                    focusVisibleClassName="Mui-focusVisible"
                    type="button"
                  >
                    <ForwardRef(ButtonBase)
                      aria-label="Download Image"
                      className="MuiButton-root MuiButton-contained makeStyles-button-7"
                      classes={
                        Object {
                          "disabled": "Mui-disabled",
                          "focusVisible": "Mui-focusVisible",
                          "root": "MuiButtonBase-root",
                        }
                      }
                      component="button"
                      disabled={false}
                      focusRipple={true}
                      focusVisibleClassName="Mui-focusVisible"
                      type="button"
                    >
                      <button
                        aria-label="Download Image"
                        className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-7"
                        disabled={false}
                        onBlur={[Function]}
                        onDragLeave={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseDown={[Function]}
                        onMouseLeave={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchMove={[Function]}
                        onTouchStart={[Function]}
                        tabIndex={0}
                        type="button"
                      >
                        <span
                          className="MuiButton-label"
                        >
                          <ForwardRef>
                            <WithStyles(ForwardRef(SvgIcon))>
                              <ForwardRef(SvgIcon)
                                classes={
                                  Object {
                                    "colorAction": "MuiSvgIcon-colorAction",
                                    "colorDisabled": "MuiSvgIcon-colorDisabled",
                                    "colorError": "MuiSvgIcon-colorError",
                                    "colorPrimary": "MuiSvgIcon-colorPrimary",
                                    "colorSecondary": "MuiSvgIcon-colorSecondary",
                                    "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                                    "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                                    "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                                    "root": "MuiSvgIcon-root",
                                  }
                                }
                              >
                                <svg
                                  aria-hidden={true}
                                  className="MuiSvgIcon-root"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M19 12v7H5v-7H3v7c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2v-7h-2zm-6 .67l2.59-2.58L17 11.5l-5 5-5-5 1.41-1.41L11 12.67V3h2z"
                                  />
                                </svg>
                              </ForwardRef(SvgIcon)>
                            </WithStyles(ForwardRef(SvgIcon))>
                          </ForwardRef>
                        </span>
                        <WithStyles(memo)
                          center={false}
                        >
                          <ForwardRef(TouchRipple)
                            center={false}
                            classes={
                              Object {
                                "child": "MuiTouchRipple-child",
                                "childLeaving": "MuiTouchRipple-childLeaving",
                                "childPulsate": "MuiTouchRipple-childPulsate",
                                "ripple": "MuiTouchRipple-ripple",
                                "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                                "rippleVisible": "MuiTouchRipple-rippleVisible",
                                "root": "MuiTouchRipple-root",
                              }
                            }
                          >
                            <span
                              className="MuiTouchRipple-root"
                            >
                              <TransitionGroup
                                childFactory={[Function]}
                                component={null}
                                exit={true}
                              />
                            </span>
                          </ForwardRef(TouchRipple)>
                        </WithStyles(memo)>
                      </button>
                    </ForwardRef(ButtonBase)>
                  </WithStyles(ForwardRef(ButtonBase))>
                </ForwardRef(Button)>
              </WithStyles(ForwardRef(Button))>
            </div>
          </div>
        </div>
      </PhotoTile>
    </SizeMeReferenceWrapper>
  </SizeMeRenderer(PhotoTile)>
</SizeMe(PhotoTile)>
`;
